### DISCUSSION TITLE
Recommendations for modelling data relationships

### DISCUSSION THREAD

**Discussion by schottra** [mswjs/msw#511] [Q&A] [answered]
I'm experimenting with using MSW as a mock layer for a REST API, and so far it's pretty neat!

One of the things I'm struggling with is the amount of boilerplate it takes to set up proper data relationships between entities. I know this is one of the points of comparison between MSW and MirageJS (with Mirage being much more geared toward creating relationships between your data).

The problem seems to be broken into two responsibilities:
1. The network/communication layer problem of mocking a server.
2. The "database" problem of properly representing your data model, relationships between entities, and reading/writing objects based on the parameters and body of a request.

MSW is solving the first problem and has no opinion on how you solve the second. So far, this has me writing my own simple database and corresponding logic to relate entities and return the right objects. It's becoming tedious and I'm imagining having to do this all over again next time I want to use MSW to create a mock server. I find myself asking, "Is it worth it for me to implement a decent portion of an Express server as a mock layer? Is this something I want to do every time I create a new project?"

Are there any tools/libraries/patterns that the authors or community are using alongside MSW to handle data modelling and relationships?

**Comment by marcosvega91** [comment]
Hi @schottra, we are working on it. We are creating a data modeling package that can be used with MSW 🥳 . Stay tuned 💯

**Reply by kettanaito** [reply]
Confirmed. Please stay tuned and expect an official recommendation to model data next year. 

Until that solution lands, you can consider libraries like [`test-data-bot`](https://github.com/jackfranklin/test-data-bot) to integrate data modeling capabilities with MSW.

**Comment by kettanaito** [comment]
A quick update on the data modeling: we are reaching the moment of a public beta release of the modeling package that's going to be called `@mswjs/data`. There are a few things left to sort out, but we hope to have it released this spring. Looking forward to you trying it out in your apps!

**Comment by kettanaito** [comment, marked as answer]
We are excited to announce the [@mswjs/data](https://github.com/mswjs/data) package. You can use that package to model data and defined relationships between the models, as well as generate request handlers for Mock Service Worker.
